From aa845a37dfc3dba82bc37cb425b3b5a6be5aa478 Mon Sep 17 00:00:00 2001
From: David Gibson <dgibson@redhat.com>
Date: Mon, 15 Sep 2014 07:12:32 -0400
Subject: [virt] kvm/ppc: fix couple of memory leaks in MPIC/XICS devices

Message-id: <1410765214-16377-17-git-send-email-dgibson@redhat.com>
Patchwork-id: 94710
O-Subject: [PATCH 16/78] KVM: PPC: fix couple of memory leaks in MPIC/XICS devices
Bugzilla: 1123145 1123133 1123367
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1123145

Signed-off-by: David Gibson <dgibson@redhat.com>

commit 458ff3c099a1266991208f2c009afc2405e5b6bc
Author: Gleb Natapov <gleb@redhat.com>
Date:   Sun Sep 1 15:53:46 2013 +0300

    KVM: PPC: fix couple of memory leaks in MPIC/XICS devices

    XICS failed to free xics structure on error path. MPIC destroy handler
    forgot to delete kvm_device structure.

    Signed-off-by: Gleb Natapov <gleb@redhat.com>
    Acked-by: Paul Mackerras <paulus@samba.org>
    Signed-off-by: Alexander Graf <agraf@suse.de>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kvm/book3s_xics.c b/arch/powerpc/kvm/book3s_xics.c
index 02a17dc..d1acd32 100644
--- a/arch/powerpc/kvm/book3s_xics.c
+++ b/arch/powerpc/kvm/book3s_xics.c
@@ -1246,8 +1246,10 @@ static int kvmppc_xics_create(struct kvm_device *dev, u32 type)
   kvm->arch.xics = xics;
  mutex_unlock(&kvm->lock);
 
- if (ret)
+ if (ret) {
+  kfree(xics);
   return ret;
+ }
 
  xics_debugfs_init(xics);
 
diff --git a/arch/powerpc/kvm/mpic.c b/arch/powerpc/kvm/mpic.c
index 2861ae9..efbd996 100644
--- a/arch/powerpc/kvm/mpic.c
+++ b/arch/powerpc/kvm/mpic.c
@@ -1635,6 +1635,7 @@ static void mpic_destroy(struct kvm_device *dev)
 
  dev->kvm->arch.mpic = NULL;
  kfree(opp);
+ kfree(dev);
 }
 
 static int mpic_set_default_irq_routing(struct openpic *opp)
-- 
1.7.1